apps/cowswap-frontend/src/modules/bridge/pure/ProtocolIcons/index.tsx

```diff
@@ -10,63 +12,215 @@ export interface ProtocolIconsProps {
   size?: number
 }
 
-const DEFAULT_ICON_SIZE = 18
-const DEFAULT_SINGLE_ICON_SIZE = 36
-const LOGO_HEIGHT_RATIO = 0.75
+const DEFAULT_ICON_SIZE = 18 // For stacked icons
+const DEFAULT_SINGLE_ICON_SIZE = 36 // For showOnlyFirst
+const LOGO_HEIGHT_RATIO = 0.5 // The ratio of the icon size to the logo height
+const STACKED_ICON_OVERLAP_RATIO = 0.4 // This drives both visual overlap and mask calculation
 
-export function ProtocolIcons({
-  secondProtocol,
-  showOnlyFirst,
-  showOnlySecond,
-  size = DEFAULT_ICON_SIZE, // Default size for stacked icons
-}: ProtocolIconsProps) {
-  // If showing only first or only second, use the specified size or default single icon size
-  const iconSize =
-    showOnlyFirst || showOnlySecond ? (size === DEFAULT_ICON_SIZE ? DEFAULT_SINGLE_ICON_SIZE : size) : size
+interface MaskConfig {
+  cx: number
+  cy: number
+  innerR: number
+  outerR: number
+}
 
-  const logoHeight = iconSize ? iconSize * LOGO_HEIGHT_RATIO : DEFAULT_ICON_SIZE * LOGO_HEIGHT_RATIO
+interface IconHoverState {
+  hoveredIcon: 'first' | 'second' | null
+  handleFirstIconMouseEnter: () => void
+  handleSecondIconMouseEnter: () => void
+  handleMouseLeave: () => void
+}
 
-  if (showOnlyFirst) {
-    return (
-      <ProtocolIcon bgColor={UI.COLOR_BLUE_900_PRIMARY} title="CoW Protocol" size={iconSize}>
-        <ProductLogo
-          variant={ProductVariant.CowProtocol}
-          height={logoHeight}
-          logoIconOnly
-          overrideColor={`var(${UI.COLOR_BLUE_300_PRIMARY})`}
-          overrideHoverColor={`var(${UI.COLOR_BLUE_300_PRIMARY})`}
-        />
-      </ProtocolIcon>
-    )
+interface MaskCalculationResult {
+  firstMaskConfig: MaskConfig | undefined
+  secondMaskConfig: MaskConfig | undefined
+  isFirstOnTop: boolean
+  isSecondOnTop: boolean
+}
+
+// Custom hook for hover state management
+function useIconHoverState(): IconHoverState {
+  const [hoveredIcon, setHoveredIcon] = useState<'first' | 'second' | null>(null)
+
+  const handleFirstIconMouseEnter = (): void => setHoveredIcon('first')
+  const handleSecondIconMouseEnter = (): void => setHoveredIcon('second')
+  const handleMouseLeave = (): void => setHoveredIcon(null)
+
+  return {
+    hoveredIcon,
+    handleFirstIconMouseEnter,
+    handleSecondIconMouseEnter,
+    handleMouseLeave,
   }
+}
 
-  if (showOnlySecond) {
-    return (
-      <ProtocolIcon title={secondProtocol.name} size={iconSize}>
-        <img src={secondProtocol.logoUrl} width={logoHeight} height={logoHeight} alt={secondProtocol.name} />
-      </ProtocolIcon>
-    )
+// Utility function for mask configuration
+const createMask = (
+  baseCx: number,
+  offsetMultiplier: number,
+  distance: number,
+  centerY: number,
+  innerR: number,
+  outerR: number,
+): MaskConfig => ({
+  cx: baseCx + offsetMultiplier * distance,
+  cy: centerY,
+  innerR: innerR,
+  outerR: outerR,
+})
+
+// Custom hook for mask calculations. Memoized only for costly computation.
+function useMaskCalculations(
```

Comment:
It's actually a pure function, not hook.
I mean, it's better to define the `useMemo` callback as a function and use it in components:
```
function calculateMask(currentDisplaySize: number, hoveredIcon: 'first' | 'second' | null) {
   const cutThickness = getBorderWidth(currentDisplaySize)
   ...
}

const { firstMaskConfig, secondMaskConfig, isFirstOnTop, isSecondOnTop } = useMemo(() =>   calculateMask(currentDisplaySize,hoveredIcon), [])
```

Reply:
Addressed.